### DISCUSSION TITLE
Copilot in Neovim is introducing syntax errors

### DISCUSSION THREAD

**Discussion by chronon** [community/community#46277] [Copilot Conversations] [answered]
### Select Topic Area

Bug

### Body

* copilot.vim 1.8.2                                                                                                                                                                 
* Neovim 0.8.3                                                                                                                                                                      
* copilot/dist/agent.js 1.71.0                                                                                                                                                      
* Node.js 18.12.1

Since v1.8.2 Copilot sometimes adds extra/duplicate characters at the end of a suggestion, creating syntax errors. Here's an example in PHP:

![copolot-1](https://user-images.githubusercontent.com/57735/216624152-24312c08-c016-423b-9ce4-158f5f3d3042.jpg)

**Comment by chronon** [comment]
Additional details: my above post is `copilot.vim` v1.8.2.

Below is v1.8.1, which does something different, but incorrect in that the existing trailing characters `);` are removed by the Copilot suggestion:

![Screen Shot 2023-02-03 at 1 26 03 PM](https://user-images.githubusercontent.com/57735/216679838-72197b0c-051b-45aa-b91d-429024a57e03.jpg)

With v1.8.0 (which I've reverted to for now) this example is working as it should:

![Screen Shot 2023-02-03 at 1 27 03 PM](https://user-images.githubusercontent.com/57735/216680007-bbf613a0-c499-42c8-ba3a-f0b1ed8dcaab.jpg)

**Comment by rphuber** [comment]
I am running into the same problem.

**Comment by chronon** [comment]
This issue does not apply to PHP only, I see the same problems with v1.8.2 in other languages (Javascript and lua confirmed).

**Comment by iamcjwilliams** [comment]
@chronon Thank you for your feedback, I will notify our copilot team

**Comment by tpope** [comment]
I cannot reproduce the issue on 1.8.2. Nevertheless I have tweaked the logic a bit, so please give the just released 1.8.3 a spin.

**Reply by rphuber** [reply]
Thanks for looking into this.  Here is a video of the issue (on the recently released 1.8.3). You can see that I don't have any auto close for `"` by the line I type on its own without copilot, but when I type out the `"` in the copilot suggestion it creates a lot of closing delimiters.  Hopefully this helps to better understand the issue.  Thanks!

https://user-images.githubusercontent.com/8245890/219389339-2e330a5f-0089-4c05-a8a4-e8fea3e9b0b4.mp4

**Reply by rphuber** [reply]
For me the issue has been around duplicate delimiters being added causing the syntax error.

**Comment by chronon** [comment]
Thanks @tpope - my previous PHP example appears to be fixed with 1.8.3. However, per the video from @rphuber and my testing with Javascript there are still some issues. Mine may be only when I have [nvim-autopairs](https://github.com/windwp/nvim-autopairs) enabled (though it works fine with 1.8.0).

#### 1.8.0 (autopairs enabled) - good
![180-autopairs](https://user-images.githubusercontent.com/57735/219396553-819fc0e0-a078-4dc6-91b8-3a894566c2ae.jpg)

#### 1.8.3 (autopairs enabled) - bad
![183-autopairs](https://user-images.githubusercontent.com/57735/219396886-ec4c66f4-e2dc-4214-b5f9-9bd91a9a1e5b.jpg)

#### 1.8.3 (autopairs disabled) - good
![183-no-autopairs](https://user-images.githubusercontent.com/57735/219397052-198b32a5-0668-4cb0-b9ef-3889fecca29a.jpg)

**Reply by rphuber** [reply]
@chronon what happens if you accept the copilot suggestion in 1.8.3, then go back and make a change say to the `100` there (with autopairs disabled)?  Do you get the same behavior as my video?

**Reply by rphuber** [reply]
it seems that the issue for me happens when the delimiter already exists and rather than using it for the suggestion completion it adds additional ones.

**Reply by chronon** [reply]
Yeah, existing delimiters seems to cause trouble. These are 1.8.3 with autopairs disabled:

![Screenshot 2023-02-16 at 10 05 22 AM](https://user-images.githubusercontent.com/57735/219403850-fc958a71-f249-4cfe-980f-05dc986af45d.jpg)
![Screenshot 2023-02-16 at 10 05 34 AM](https://user-images.githubusercontent.com/57735/219403910-8e642a71-d77b-4653-9c12-ed5596459257.jpg)

**Comment by akhilsbehl** [comment]
Hi, I've this issue in Python as well. How can I help to solve this?

**Comment by rphuber** [comment]
@iamcjwilliams @tpope any update here?  I tried to give context around what I feel is happening and it's still happening.  Thanks

**Comment by chronon** [comment]
@tpope (cc: @iamcjwilliams) - An update on this would be appreciated since we haven't seen a copilot.vim release since Feb 15th, the issue still exists for me and others, and Copilot is a paid product. Thanks!

**Comment by chronon** [comment]
@rphuber and anyone else having problems per this discussion with `copilot.vim` (after v1.8.0) - the unofficial [pure lua replacement](https://github.com/zbirenbaum/copilot.lua) is working great for me so far. Besides **not** creating syntax errors, it works fine with auto-pairs, and has the added bonus of optional keymaps to accept suggestions by word and/or line which I'm finding very useful.

**Comment by tpope** [comment, marked as answer]
This should be fixed (finally) in Copilot.vim 1.8.4.
